### ISSUE TITLE
LSP speed improvements

### ISSUE THREAD

**Issue by anjmao** [saibing/bingo#22]
Currently bingo is doing full file content replacement on each change https://github.com/saibing/bingo/blob/master/langserver/internal/source/file.go#L55 and removing the whole ast tree, token and pkg. After that completion request need's to parse the whole file again which is no efficient in a lot of cases.

Simplest example
```go
package main

import "fmt"

func main() {
   fmt.Println("hello")
   fmt.#
}
```
Few solutions:
1. If I want to get autocomplete for fmt. statment there is no need to fully remove whole file ast, instead it should only add needed new nodes to existing ast tree. See [astutil package](https://github.com/golang/tools/blob/master/go/ast/astutil/rewrite.go)
2. Debounce input and implement cancelation. If I quickly type fmt I will get 3 requests to lang server. Debounce can reduce overhead a lot.

There is a lot more cases then file ast should not be fully removed. Issues https://github.com/saibing/bingo/issues/21 and https://github.com/saibing/bingo/issues/14 are mostly occurs due to this issue.

At the moment I'm investigating how Typescript LSP is handling file changes since it works super fast. If you look at https://github.com/Microsoft/TypeScript/blob/master/src/services/textChanges.ts there is tons of cases for such performance improvements.

**Comment by saibing** [comment]
@anjmao 

thank you for your feedback.

I have improved the performance of bingo's LSP, please download the latest version bingo to try it out.

I have added a cache layer for [github.com/golang/tools/go/packages](https://github.com/saibing/tools/commit/acb945d26e2167673087885042cca145f5189773)

**Comment by oblitum** [comment]
@saibing I'm currently getting the following when building bingo:

```
# github.com/saibing/bingo/langserver/internal/cache
langserver/internal/cache/global_cache.go:56:2: undefined: packages.NoUseCache
langserver/internal/cache/global_cache.go:57:2: undefined: packages.DebugCache
```

**Comment by saibing** [comment]
@saibing 

bingo is a go module project, so you cannot install it via to 'go get github.com/saibing/bingo".

you need do as follow:
```bash
git clone https://github.com/saibing/bingo.git
cd bingo
GO111MODULE=on; go build
```

**Comment by oblitum** [comment]
OK thanks! Got that error fixed.

**Comment by seagle0128** [comment]
I  got this error now:

```shell
# github.com/saibing/bingo/langserver/internal/cache
langserver/internal/cache/project.go:55:2: undefined: packages.DebugCache
langserver/internal/cache/project.go:56:2: undefined: packages.ParseFileTrace
langserver/internal/cache/project.go:269:22: p.view.Config.Cache undefined (type *packages.Config has no field or method Cache)
langserver/internal/cache/project.go:273:22: p.view.Config.Cache undefined (type *packages.Config has no field or method Cache)
langserver/internal/cache/project.go:328:35: undefined: packages.WalkFunc
langserver/internal/cache/project.go:335:22: p.view.Config.Cache undefined (type *packages.Config has no field or method Cache)
langserver/internal/cache/view.go:37:4: unknown field 'Cache' in struct literal of type packages.Config
```

My command:
```shell
git clone https://github.com/saibing/bingo.git
cd bingo
GO111MODULE=on; go install
```
or 
```shell
GO111MODULE=on; go get -u github.com/saibing/bingo
```

**Comment by anjmao** [comment]
@saibing I have few ideas how implement that, I hope I will manage to make PR to bingo this week :)

**Comment by anjmao** [comment]
Here is some work I done today. My idea is:
1) Get whole line content.
2) Use `parse.ParseExpr` to convert that line to ast nodes. If it fails to parse just skip it.
3) Check if there is already BlockStmt node on the same line. This is important since we need at least one node on that line to preserve other tokens Pos.
4) If 3 is true, replace nodes. The most tricky part here is to map correctly and compute missing Pos for new ast nodes since ParseExpr doesn't return correct Pos.

I was able to get current line ast nodes, here is unit tests in which I'm experimenting on https://github.com/anjmao/bingo/blob/ast-node-replace/langserver/internal/cache/view_test.go#L41
The only remaining logic is to replace actual file ast nodes.

**Note:**
For expression `fmt.Println("hi")` to type we need 17 characters of which only `fmt.` and `fmt.Println(` and not parsable by ParseExpr func which sounds quite promising since we are still saving a lot.

@saibing 
Let me know what do you think, thanks.

**Comment by anjmao** [comment]
After trying astutil.Apply to rewrite ast I found few issues.

1. ParseExpr doesn't work well with astutil.Apply, the recommended solution was to construct new ast manually which may be ok.
2. token.File info becomes out of sync. Token line, lineInfos and size may b incorrect which affects autocomplete, find references etc.

Here is the issue details https://github.com/golang/go/issues/29615

**Comment by s-kostyaev** [comment]
After latest commits it works slowly. I think it was after remove `golist-duration` flag.

**Comment by inliquid** [comment]
@s-kostyaev I came to same conclusion, without golist cache (it was removed completely) its speed has degraded.

**Comment by saibing** [comment]
@s-kostyaev @inliquid 

You are correct, running 'go list' command will cause the completion to slow down, especially on Windows. the go list cache will significantly increase the speed of completion, but unfortunately this will lead to various incorrect results. So I deleted it, I am still thinking about other methods.

**Comment by s-kostyaev** [comment]
@saibing on windows and on mac `fork` is very slow. So you can try to execute `go list` methods instead of use `os.Exec` it can improve speed incredibly.

**Comment by saibing** [comment]
@anjmao @s-kostyaev @inliquid 

This problem has been basically solved in the commit [31f82b6aa7367caf3a492e9550d92346c305be21](https://github.com/saibing/bingo/commit/31f82b6aa7367caf3a492e9550d92346c305be21)
